src/nn/generalized_residual.rs

```diff
@@ -0,0 +1,379 @@
+use crate::prelude::*;
+
+/// A residual connection `R` around `F`: `F(x) + R(x)`,
+/// as introduced in [Deep Residual Learning for Image Recognition](https://arxiv.org/abs/1512.03385).
+///
+/// # Generics
+/// - `F`: The underlying module to do a skip connection around.
+/// - `R`: The underlying residual module
+///
+/// # Examples
+/// ```rust
+/// # use dfdx::prelude::*;
+/// let module: Residual<ReLU, Square> = Default::default();
+/// let x = Tensor1D::new([-2.0, -1.0, 0.0, 1.0, 2.0]);
+/// let y = module.forward(x);
+/// assert_eq!(y.data(), &[4.0, 1.0, 0.0, 2.0, 6.0]);
+/// ```
+#[derive(Debug, Clone, Default)]
+pub struct GeneralizedResidual<F, R>(F, R);
+
+impl<F: CanUpdateWithGradients, R: CanUpdateWithGradients> CanUpdateWithGradients
+    for GeneralizedResidual<F, R>
+{
+    /// Pass through to `F`'s [CanUpdateWithGradients].
+    fn update<G: GradientProvider>(&mut self, grads: &mut G) {
+        self.0.update(grads);
+        self.1.update(grads);
+    }
+}
+
+impl<F: ResetParams, R: ResetParams> ResetParams for GeneralizedResidual<F, R> {
+    /// Pass through to `F`'s [ResetParams].
+    fn reset_params<RNG: rand::Rng>(&mut self, rng: &mut RNG) {
+        self.0.reset_params(rng);
+        self.1.reset_params(rng);
+    }
+}
+
+impl<F, R, T, O> Module<T> for GeneralizedResidual<F, R>
+where
+    T: Tensor<Dtype = f32>,
+    O: Tensor<Dtype = T::Dtype, Tape = T::Tape>,
+    F: Module<T, Output = O>,
+    R: Module<T, Output = O>,
+{
+    type Output = O;
+
+    /// Calls forward on `F` and `R` and then sums their result: `F(x) + R(x)`
+    fn forward(&self, x: T) -> Self::Output {
+        let (x, tape) = x.split_tape();
+
+        // do R(x) on the tape
+        let (r_x, tape) = self.1.forward(x.duplicate().put_tape(tape)).split_tape();
+
+        // do F(x) on the tape
+        let f_x = self.0.forward(x.put_tape(tape));
+
+        add(f_x, &r_x)
+    }
+}
+
+impl<F: SaveToNpz, R: SaveToNpz> SaveToNpz for GeneralizedResidual<F, R> {
+    /// Pass through to `F`/`R`'s [SaveToNpz].
+    fn write<W>(
+        &self,
+        filename_prefix: &str,
+        w: &mut zip::ZipWriter<W>,
+    ) -> zip::result::ZipResult<()>
+    where
+        W: std::io::Write + std::io::Seek,
+    {
+        // I have no idea what to put here
+        self.0.write(filename_prefix, w)?;
+        self.1.write(filename_prefix, w)?;
```

Comment:
I think these are why the save/load is failing :) basically the first parameter is the prefix of the file to write the data to. Since they both are using `filename_prefix`, they are overwriting each other!

You can use something like this for the first param of both lines: `&format!("{filename_prefix}0.")` and `&format!("{filename_prefix}1.")`, which will save the first param to `0.<submodule>.npy` and `1.<submodule>.npy`.

Same comment for LoadFromNpz

Reply:
thanks for the help! it works now